python/mxnet/kvstore/kvstore.py

```diff
@@ -104,8 +104,22 @@ def broadcast(self, key, value, out, priority=0):
         [ 2.  2.  2.]]
 
         """
-        self.init(key, value)
-        self.pull(key, out=out, priority=priority)
+        cvkeys, cvals, use_str_keys = _ctype_key_value(key, value)
+        if out is not None:
+            cokeys, couts, _ = _ctype_key_value(key, out)
+        else:
+            cokeys = cvkeys
+            couts = cvals
+
+        if use_str_keys:
+            check_call(_LIB.MXKVStoreBroadcastEx(
+                self.handle, mx_uint(len(cvkeys)), cvkeys, mx_uint(len(cokeys)), cokeys,
+                cvals, couts, ctypes.c_int(priority)))
+        else:
+            check_call(_LIB.MXKVStoreBroadcast(
+                self.handle, mx_uint(len(cvkeys)), cvkeys, mx_uint(len(cokeys)), cokeys,
+                cvals, couts, ctypes.c_int(priority)))
+
 
```

Comment:
Shall we have p3store.py that override `is_capable`, `set_optimizer`, etc?

Reply:
Shall we also add tests to https://github.com/apache/incubator-mxnet/blob/d9291e5ae0b3f757e781bf457bdff8b6a61b8c8d/ci/docker/runtime_functions.sh#L1373

Reply:
the register(klass) function uses klass.__name__ as the name of the kvstore. So I guess we need to register multiple classes for dist_p3, dist_device_p3. On the other hand, if we are going with my suggestion in the previous comment, we won't even be having a name for the new kvstore.

I'll get back to this once we finalize what to do with the previous issue.

Reply:
I'll write the test cases, once we finalize the design. otherwise I probably have to rewrite the tests  multiple times. I hope that is okay.

Reply:
What about adding a name flag to `register`? I had a quick test:
```
import mxnet as mx
import warnings
from mxnet.kvstore import KVStoreBase

@staticmethod
def register(klass, name=None):
    assert(isinstance(klass, type))
    if name is None:
        name = klass.__name__.lower()
    else:
       name = name.lower()
    if name in KVStoreBase.kv_registry:
        warnings.warn('WARNING: New kvstore %s.%s is overriding '
                      'existing kvstore %s.%s' %
                      (klass.__module__, klass.__name__,
                       KVStoreBase.kv_registry[name].__module__,
                       KVStoreBase.kv_registry[name].__name__))
    KVStoreBase.kv_registry[name] = klass
    return klass

mx.kvstore.KVStoreBase.register = register
class MyClass:
    pass

# manually call register multiple times 
mx.kvstore.KVStoreBase.register(MyClass, name='class_dist')
mx.kvstore.KVStoreBase.register(MyClass, name='class_dist_device')

print(KVStoreBase.kv_registry)
```

returns 
```
{'teststore': <class 'mxnet.kvstore.base.TestStore'>, 
'class_dist': <class '__main__.MyClass'>, 
'class_dist_device': <class '__main__.MyClass'>}
```

Reply:
if a user want to directly use kvstore class, then probably have to use dist_p3 to instantiate kvstore python class and also set DMLC_PS_VAN_TYPE env var. seems a bit redundant to me.

I feel that we don't need separate python class for p3 except for overriding is_capable(). But I think if we are planning to use DMLC_PS_VAN_TYPE in the backend anyway, we probably can raise exception in the backend if someone tries to set optimizer or updater in kvstore. So we don't need any changes in the python frontend at all.

Reply:
Sounds good. We don't need to have separate class then. We are already raising exceptions in the backend when setting optimizer for P3. What about also changing `is_capable` in KVStore to return the value based on `DMLC_PS_VAN_TYPE`?

Reply:
yes. sounds good. I'll update the code accordingly